DEV-GUIDELINES.md

```diff
@@ -0,0 +1,187 @@
+3. We only use java 8 source compatibility (with the sole exception of module-info.java mentioned above)
+    1. See useful helpers in common that provide some of the new nice features backported to java 8
+    
+# Testing
+
+We use JUnit 5 with Hamcrest assertions.
+
+The Hamcrest assertion API differs a lot from JUnit assertion API and when both are used,
+ the tests are hard to read.
+ 
+We have chosen the Hamcrest approach:
+- assertThat(actualValue, assertion(expectedValue))
+
+Main import: `import static org.hamcrest.MatcherAssert.assertThat;` - the main method to be used
+
+Most commonly used assertions are available in `org.hamcrest.CoreMatchers`:
+1. is() - assertion of equality: `assertThat(value, is(true)))`
+2. notNullValue() - assertion of a non-null value: `assertThat(value, notNullValue())`
+3. nullValue() - assertion of null value: `assertThat(value, nullValue())`
+4. startsWith(String) - assertion that string is prefixed: `assertThat(value, startWith("a"))`
+5. endsWith(String) - assertion that string is suffixed: `assertThat(value, endsWith("b"))`
+
+For other nice assertions see the CoreMatchers class or google ;)
+
+The following assertions may be used from JUnit 5:
+1. `import static org.junit.jupiter.api.Assertions.assertAll;` - doing multiple assertions at once (more than one may fail)
+2. `import static org.junit.jupiter.api.Assertions.assertThrows;` - asserting that an expression throws an exception
+
+
+**Example of why we have this rule:**
+
+Original assertion:
+```java
+assertTrue(ex.getMessage().contains("'" + config.key() + "'"));
+```
+The output of the test was:
+`Expected: true, actual: false`
+
+Refactored assertion:
+```java
+assertThat(ex.getMessage(), containsString("'" + config.key() + "'"));
+```
+
+New output:
+```
+Expected: a string containing "'list-1'"
+but: was "Requested value for configuration key 'list-1.1' is not present in the configuration."
+```
+
+# Maven
+1. All third party versions are managed in root pom
+    1. Plugins
+    2. Dependencies
+2. Adding a new third party dependency (or upgrading to a newer version) requires an 
+    internal process to be carried out. In such cases a delay is to be expected when merging.
+3. When referencing other Helidon modules, use ${project.version} as a version
+4. In pom.xml of a module, always define a "name" tag and follow naming conventions already used
+    1. For project module: Helidon ${module_name} Project
+    2. For java module: Helidon ${module_name} 
+5. All java modules that are expected to be used by our users MUST be defined in our [bom pom](bom/pom.xml)
+6. Bundles may be created, though we still must give our users the freedom to pick and choose modules directly  
+7. Java EE components and Microprofile specifications should be in "provided" scope unless you are implementing
+    the spec itself
+8. Carefully choose scope for dependencies on other helidon modules (e.g. microprofile extensions should have
```

Comment:
The problem I was pointing to is that we have sometimes scope "compile" where we really should have "provided" or "optional"
Part of this problem can be seen when running with jigsaw and java 11, as that shows all the places where we use a dependency from different locations or when we depend on an API in compile scope and another component is implementing it...

Reply:
I understood your point. Since "compile" is the default scope, forcing the scope to be explicitly declared would help limit this sort of problems (visible and could be caught during review)

Reply:
Agreed to leave default scope unfilled.